### PR TITLE
Ensure filter blocks are not reloaded every time they are selected in the editor

### LINKED ISSUE #7676: Filter control blocks in the editor reload every time they are selected
## Describe the bug

Every time the merchant selects the _Filter by Price Controls_, _Filter by Attribute Controls_, _Filter by Stock Controls_ and _Filter by Rating Controls_ in the editor, it causes the block to go into a loading state (displaying the placeholders) unnecessarily.

## To reproduce

1. Add the Filter by Price block to a post or page.
2. Click on the Filter by Price Controls block.
4. Notice it shows the loading state for a few seconds.
5. Click on the Filter by Price title and back into the controls block. Notice every time you select the controls block, it shows the loading state.

## Expected behavior

From a user perspective, there is no need for the block to go into a loading state when selected.

## Screenshots

[Enregistrament de pantalla des de 14-11-22 13:32:44.webm](https://user-images.githubusercontent.com/3616980/201661118-2e91c6c6-9289-49f2-9727-be763ea8b590.webm)

### PR DESCRIPTION
**Pull request by nefeline**
Fixes #7676

This update ensures the filter components are not reloaded every time they are selected in the editor.

### Why was this problem happening in the first place?

On the `useCollectionData` hook, [we are debouncing](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/context/hooks/collections/use-collection-data.ts#L148-L153) `shouldSelect` to delay data collection requests on the initial mount for 200ms. By reading the original PR where it was introduced, #1233, this was added to prevent multiple collection data requests, which makes total sense.

The problem here is when `shouldSelect` is false, the `useCollection` hook [returns null](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/context/hooks/collections/use-collection.ts#L80-L82). In the editor, whenever we click on a component, it is re-rendered. Since the initial request is debounced (with `shouldSelect` being `false` by default), we always end up fetching again from the store.

To circumvent this, the `isEditor` and `isSelected` props are now passed to the `useCollectionData` and `useCollection` hooks as params. If we are in the editor, `shouldSelect` is set to `true` by default. This way, we will always fetch the latest data when the editor page is initially loaded.

When a given block is selected in the editor, we now return the same data from the initial request (from `currentResults`) instead of fetching from the store again on re-render. This change only affects the editor, where the components [are disabled](https://developer.wordpress.org/block-editor/reference-guides/components/disabled/) by default.

### Screenshots

| Before | After |
| ------ | ----- |
|  |  |

#### User Facing Testing

1. Create a new post, add the following blocks: Filter by rating, filter by stock, filter by attribute, filter by price, Products (Beta), and publish it.
3. Access the post as a regular customer would. Ensure nothing has changed, and all filters are working as expected.
4. Now edit the post: ensure all components are normally loaded in the editor. When you click on them, ensure that you see the same behavior as demonstrated on [this gif](https://user-images.githubusercontent.com/15730971/208894840-5fed09d6-2c5c-4da2-9ee3-370068709309.gif) shared earlier on this PR (no preloaders should be displayed).

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

We have some improvements in performance here, since, now, in the editor, we are only fetching data from the store on the initial render, relying on the `currentResults` whenever the component is selected.

### Changelog

> The filter by attribute, price, rating, and stock blocks are not reloaded when selected in the editor anymore.

### CODE REVIEW

**Review by dinhtungdu — COMMENTED** [review]
Thank you so much for the PR @nefeline! Great work, I can confirm that the filter block is no longer flashing when clicking on them.

> On the `useCollectionData` hook, [we are debouncing](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/context/hooks/collections/use-collection-data.ts#L148-L153) `shouldSelect` to delay data collection requests on the initial mount for 200ms. By reading the original PR where it was introduced, #1233, this was added to prevent multiple collection data requests, which makes total sense.

In testing this, I realize that there are still multiple collections data requests made on the page that contain multiple PR blocks. Should we revisit 1233?

<details>
<summary>Screenshot</summary>

image

</details>

We're passing `isEditor` and `isSelected` to our `useCollection*` hooks, but as we're making the `debouncedShouldSelect` not 'debounced' in the editor, I think only passing `isEditor` is enough, as `shouldSelect` is now always `true` in the editor, so the collection hook won't return `null` anymore. What do you think?

**Comment by nefeline** [comment]
> Thank you so much for the PR @nefeline! Great work, I can confirm that the filter block is no longer flashing when clicking on them.

Thanks 🙌 !

> In testing this, I realize that there are still multiple collections data requests made on the page that contain multiple PR blocks. Should we revisit 1233?

Yes, I agree there's room for improvements regarding 1233; from my understanding, the debounce was added as a workaround to reduce the number of API requests, due to the absence of a shared context, which makes sense: if we can structure the codebase in a way we know for a fact that x filter components exist on a given post/page and make only one request based on this shared information, that would be the ideal. If we stick with the `debounce` path, for now, there are other avenues I believe we can explore, such as increasing the debounce value and comparing the `currentQuery` with the `query` and returning the `currentResult` if they are the same as well (open to hearing your thoughts if you have other ideas here!). I'd say this is a bit bigger than the scope of this PR, which targets the editor exclusively and aims to prevent reload when filter components are selected in this area. Perhaps opening a separate PR to address this in the customer-facing area could be a good path to follow next? What do you think?

 > We're passing isEditor and isSelected to our useCollection* hooks, but as we're making the debouncedShouldSelect not 'debounced' in the editor, I think only passing isEditor is enough, as shouldSelect is now always true in the editor, so the collection hook won't return null anymore. What do you think?

`isSelected` was added here to be super conservative, as I wanted to isolate the changes to the scope of the selected blocks and not all blocks, but in retrospect, it might make sense to have the same behavior across all filter components in the editor. I'll go ahead & review/update this 👍.

File: assets/js/base/context/hooks/collections/use-collection.ts [hunk 1]
```diff
@@ -77,9 +79,23 @@ export const useCollection = (
 	const throwError = useThrowError();
 	const results = useSelect(
 		( select ) => {
+			if (
+				isEditor &&
+				typeof currentResults === 'object' &&
+				currentResults.hasOwnProperty( 'current' ) &&
+				currentResults.current.hasOwnProperty( 'results' ) &&
```
Comment (dinhtungdu) [inline assets/js/base/context/hooks/collections/use-collection.ts]: We can use `isObject` from `@woocommerce/types` here to make this call shorter, something like below. It also fixes the current TS error:

image


```suggestion
				isObject( currentResults?.current?.results ) &&
```

**Review by dinhtungdu — APPROVED** [review]
Thanks so much for the update @nefeline! This is testing great on my end. I left an optional comment but I'm pre-approving this. 🚀

File: assets/js/base/context/hooks/collections/use-collection.ts (hunk 1 above)
Comment (nefeline) [reply assets/js/base/context/hooks/collections/use-collection.ts]: Good call, updated!

**Pull request merged**